src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,379 @@
+                var analysisContext = buildersToExpressionContext[builderNode];
+                rewrittenBuildersNode = analysisContext.Node.RewrittenExpression;
+                constantsMapper = analysisContext.Node.ConstantsRemapper;
+                argumentTypeName = analysisContext.Node.ArgumentTypeName;
+            }
+            nodesRemapping.Add(builderNode, rewrittenBuildersNode);
+            constantMappers.Add(constantsMapper);
+        }
+
+        SyntaxNode buildersExpression = null;
+        if(RHS is IdentifierNameSyntax identifierNameSyntax)
+        {
+            buildersExpression = variableValues[identifierNameSyntax.Identifier.ValueText].Node.RewrittenExpression;
+        }
+        else if(RHS is InvocationExpressionSyntax invocationExpression)
+        {
+            buildersExpression = buildersToExpressionContext[invocationExpression].Node.RewrittenExpression;
+        }
+        else
+        {
+            buildersExpression = RHS.ReplaceNodes(nodesRemapping.Keys, (n, _) => nodesRemapping[n]);
+        }
+
+        if (syntaxKind == SyntaxKind.AndAssignmentExpression ||
+            syntaxKind == SyntaxKind.OrAssignmentExpression)
+        {
+            var identifier = variableNames.FirstOrDefault();
+            var isValidVariable = variableValues.ContainsKey(identifier);
+            if (!isValidVariable) { return; }
+            var LHS = variableValues[identifier].Node.RewrittenExpression;
+            if (syntaxKind == SyntaxKind.AndAssignmentExpression)
+            {
+                syntaxKind = SyntaxKind.BitwiseAndExpression;
+            }
+            else if (syntaxKind == SyntaxKind.OrAssignmentExpression)
+            {
+                syntaxKind = SyntaxKind.BitwiseOrExpression;
+            }
+
+            buildersExpression = SyntaxFactory.BinaryExpression(syntaxKind, LHS as ExpressionSyntax, buildersExpression as ExpressionSyntax);
+        }
+        
+        var (rewrittenBuildersExpression, compoundConstantsMapper) = ConstantsMapperComposer.Compose(buildersExpression, constantMappers);
+        ExpressionAnalysisContext variableInfo = new ExpressionAnalysisContext(new ExpressionAnalysisNode(RHS, argumentTypeName, rewrittenBuildersExpression, compoundConstantsMapper));
+        foreach (var variableName in variableNames)
+        {
+            dirtyVariables.Add(variableName);
+            if (!variableValues.ContainsKey(variableName))
+            {
+                variableValues.Add(variableName, variableInfo);
+            }
+            else
+            {
+                variableValues[variableName] = variableInfo;
+            }
+        }
+    }
+
+    private static (SyntaxNode statementNode, bool canEvaluate, HashSet<string> dirtyVariables) HandleDeclaration(VariableDeclaratorSyntax variableDeclarator, NodeContext nodeContext, BuildersAnalysis buildersAnalysis)
+    {
+        SyntaxNode statementNode = variableDeclarator;
+        var buildersNodes = new List<SyntaxNode>();
```

Comment:
Try minimizing creation of new objects in case it's not used.
For example buildersNodes are acutalluy needed only when canEvaluate == true.

Also should builderNodes be renamed to childBuilderNodes.

Reply:
Done